src/core/processing/qgsprocessingregistry.h

```diff
@@ -138,11 +175,29 @@ class CORE_EXPORT QgsProcessingRegistry : public QObject
     //! Emitted when a provider is removed from the registry
     void providerRemoved( const QString &id );
 
+    /**
+     * Emitted when a new parameter type has been added to the registry.
+     *
+     * \since QGIS 3.2
+     */
+    void parameterTypeAdded( QgsProcessingParameterType *type );
+
+    /**
+     * Emitted when a parameter type has been removed from the
+     * registry and is about to be deleted.
+     *
+     * \since QGIS 3.2
+     */
+    void parameterTypeRemoved( QgsProcessingParameterType *type );
+
   private:
 
     //! Map of available providers by id. This class owns the pointers
     QMap<QString, QgsProcessingProvider *> mProviders;
 
+    //! Hash of available parameter types by id. This object owns the pointers.
+    QHash<QString, QgsProcessingParameterType *> mParameterTypes;
```

Comment:
My understanding is that QMap would actually be faster here than QHash -- see https://woboq.com/blog/qmap_qhash_benchmark.html. Conclusions there are that for small containers (<32 elements) QMap is faster, and also that QMap is often faster anyway for qstring based keys.